DRAFT: pulling in ADQL 2.1 support

tsdower opened the pull request.

gmantele committed:
[ADQL,TAP] New parser for ADQL-2.1.

gmantele committed:
[ADQL] Do not support anymore the function REGION in the ADQL-2.1 parser.

This function has indeed be completely removed from the ADQL standard (v2.1).

gmantele committed:
[ADQL] Add a mechanism to declare ADQL features and to set which optional ones
are supported or not. All ADQLObject instances now have a new function allowing
to return a LanguageFeature describing the corresponding ADQL feature.

Only all geometric functions (e.g. POINT, CONTAINS) and the new ADQL-2.1
function `LOWER` are optional. By default, the ADQL-2.1 parser declare them as
supported. But their support can be changed at anytime thanks
ADQLParser.getSupportedFeatures() and ADQLParser.setSupportedFeatures(...).

In the current state, UDF declaration does not work anymore with the ADQL-2.1
parser. This will be fixed in a further commit. Besides, TAP has not yet been
adapted to use/configure the FeatureSet of the parser to use.

Anyway, the Javadoc of all geometric functions has been updated (so that being
conform to their description in ADQL-2.1) as well as the formatting of all
ADQLObjects (e.g. lines fitting on 80 characters, ...).

gmantele committed:
[ADQL] Update the README with the correct ADQL version.

gmantele committed:
[ALL] Move the license files at the root of the GitHub repository.

gmantele committed:
[ADQL] Fix the language feature name of INTERSECTS.

gmantele committed:
[ADQL] Add the version of the used ADQL grammar in ADQLQuery + Make geometries
optional features of the ADQL-2.0 grammar (thus, supported geometries must be
declared as in the ADQL-2.1 parser: with the FeatureSet of the parser).

TODO: see how to throw errors when un-supported geometries are used in STC-S
      as done before at the DBChecker level.

gmantele committed:
[ADQL] Add an option to FeatureSet to allow/forbid non-declared UDFs.

gmantele committed:
[ADQL] `FeatureSet.unsupportAll()` must un-support all supported features
instead of ONLY those listed in `availableFeatures`.

gmantele committed:
[ADQL] Fix the way to test UDF features in FeatureSet
(i.e. `Object.equals` instead of `==`).

gmantele committed:
[ADQL] Make UDF optional features. A special LanguageFeature constructor must be
used: `LanguageFeature(FunctionDef)`. This way, the syntax of the function
definition string is checked and function name and parameters can be correctly
extracted. These information are particularly important to determine at the
ADQLParser level whether a DefaultUDF is supported or not while not knowing the
parameters and function types.

gmantele committed:
[ADQL] Check the UDF name while creating a FunctionDef and so, also while
creating the corresponding LanguageFeature. Now, FunctionDef throws a
ParseException when the UDF name is invalid.

gmantele committed:
[ADQL] Add an easy UDF translation. An SQL expression template can now be
provided. When done, all $$i (where i is the index of a parameter ; starting
from 1) are replaced by the SQL translation of the corresponding parameters.

gmantele committed:
[ADQL] No more need to remove FunctionDef with empty names...now, by definition,
this is not any more possible.

gmantele committed:
[ADQL] ADQLParser is no longer a Java Interface and ADQLParserFactory does not
exist anymore.

ADQLParser is now a mix between the former ADQLParserFactory and ADQLParser2xx.
All ADQLParser2xx resulting from the compilation of the JavaCC files are
now named ADQLGrammar2xx and implement the new interface ADQLGrammar. The JavaCC
parsers are no longer used directly. This is the role of ADQLParser to simplify
the parser usage and to allow an easy switch between different grammar versions.

Besides, for more clarity in the class organisation, all generated parser
classes have been moved into the package `adql.parser.grammar`.

gmantele committed:
[ADQL] Add examples of ADQL parsing.

gmantele committed:
[ADQL] Move validation of coord. sys. and STC-s expressions from DBChecker to
ADQLParser.

gmantele committed:
[ADQL] Add ILIKE

gmantele committed:
[ADQL] Make the coordinate system argument optional in geometries constructor
(e.g. POINT, BOX, CIRCLE and POLYGON).

gmantele committed:
[ADQL] Efficient PgSphere translation of the Preferred X-Match Syntax described
in the ADQL-2.1 standard (in section 4.2.7).

Currently, geometries are not translated in MySQL and MS-SQL Server translators.
So, the preferred xmatch syntax has not been implemented in these translators.

gmantele committed:
[ADQL] Support DISTANCE also with 4 parameters (i.e. lon1, lat1, lon2, lat2).

gmantele committed:
[ADQL] Support OFFSET.

gmantele committed:
[ADQL] Adapt the previous commit to add OFFSET into the FeatureSet. With the
previous commit, OFFSET was always available although it is actually an optional
feature.

gmantele committed:
[ADQL] Support hexadecimal values (as numeric values).

gmantele committed:
[ADQL] Support bitwise operators + Fix wrong hexadecimal conversion (only for
negative values).

The implemented operators precedence is:

```
[~-](unary) >> [*/] >> [+-] >> [^&|]
```

_**Example:** `~3-1|2*5^6/1+2 = ((~3)-1)|((2*5)^((6/1)+2)) = -5`_

gmantele committed:
[ADQL] Support the function IN_UNIT(...). (just for parsing....no translation
done because the function must also be available in the database)

gmantele committed:
[ADQL] Improve error report in case of incorrect argument for numeric functions.

gmantele committed:
[ADQL] Add default translation for IN_UNIT(...): as geometries, same as in ADQL.

gmantele committed:
[ADQL] An ADQLTranslator now provides a list of all features it supports,
& Rename UnitConversionFunction into InUnitFunction.

gmantele committed:
[ADQL] Fix MOD, RAND, ROUND and TRUNCATE (see ADQL-2.1 and ADQL-2.0 Erratum 2).

gmantele committed:
[ADQL] 4 commits in one: 1/ new syntax of ORDER and GROUP BY, 2/ fix positions
of ORDER items, 3/ improve error messages in case of incorrect function argument
(especially in WHERE), and 4/ run JUnit tests for all ADQL versions (only in
TestDBChecker and TestADQLParser).

gmantele committed:
[ADQL,TAP] + isCaseSensitive() in DBSchema, DBTable and DBColumn + Change rules
about TAP_SCHEMA columns `schema_name`, `table_name` and `column_name`
(see tap_schema_requirements.md for more details).

gmantele committed:
Support WITH + fix a lot the case-sensitivity management. WARNING: the signature
of the constructors of DefaultDBColumn and DefaultDBTable has changed!

gmantele committed:
[ADQL] Add a Unicode character (\uF0A0) as space in QueryFixer.

gmantele committed:
[ADQL] Remove bitwise operators and hexadecimal values....they are not yet
mature as defined in PR-ADQL-2.1 (see slides 26 of the talk
["Feedback on the implementation of ADQL-2.1"](https://wiki.ivoa.net/internal/IVOA/InterOpOct2019DAL/feedback_adql2_1.pdf)).

gmantele committed:
[ADQL] Remove column definition list in `WITH` clause (according to the recent
modifications of ADQL-2.1-PR)
+ Complete and fix case-sensitivity tests.

gmantele committed:
[ADQL] Allow also `CENTROID` and UDF as arguments of `DISTANCE` (2-arg form)
& Allow UDF as valid geometry argument (a UDF being able to return geometries)
like in functions `AREA`, `CENTROID`, `DISTANCE`, ...

gmantele committed:
[ALL] Upgrade Java version (8+) + Upgrade Gradle configuration

gmantele committed:
[ADQL] Apply "ADQL-2.0 Erratum-3" :

The now optional coordinate system argument of `POINT`, `BOX`, `CIRCLE` and
`POLYGON` MUST be a _string literal_.

_See <https://wiki.ivoa.net/twiki/bin/view/IVOA/ADQL-2_0-Err-3>_

gmantele committed:
[ADQL] BOX+CIRCLE+POLYGON : make coord. sys. optional and allow POINT values

gmantele committed:
[ALL] Update Gradle to work in english (especially when running JUnit tests)

gmantele committed:
[ADQL] Fix incomplete conflict resolution

gmantele committed:
[ADQL] Fix OFFSET translation for MS-SQLServer.

gmantele committed:
[ADQL] Add `UPPER(...)`

gmantele committed:
[ADQL] Restore `REGION(...)` & Add `CAST(...)` & JavaCC-7.0.10 in Gradle

This time `REGION(...)` can only accept string literal as single parameter and
only the DALI and the STC/s serializations are allowed. It is however possible
to accept any string expression and/or allow more serializations ; to do that,
`ADQLParser.allowExtendedRegionParam(true)`.

In the current state `CAST(...)` supports only the standard datatypes and no
geometrical one. In a further version, it may be possible to specify custom
datatypes and how to translate/cast them.

The JavaCC's grammar files are now compiled with JavaCC-7.0.10 (released in
Nov. 2020). Generated files like ParseException, TokenManager, etc have been
upgraded too, while keeping the customization required by the VOLLT/ADQLLib.
The generation of JavaCC files is now possible with Gradle through the tasks
`build` and `javacc`.

Also fixed:

- Custom type creation while parsing an unknown type in the UDF definition
  (FunctionDef.parse) : the length attribute was included in the type's name
  instead of being apart (ex: `custom(10)` was interpreted as the custom type
  named `custom(10)` instead of `custom` with the length '10').

- Fix error message when a MathFunction has more parameter as expected.

gmantele committed:
[ADQL] Fix parsing of `DOUBLE PRECISION`: support space characters in between.

gmantele committed:
[ADQL] Allow CAST(...) in expressions and check the resulting datatype in its
context.

gmantele committed:
[ADQL] Improve translation pattern with the syntax `$i..` and the ternary
conditional expression (`$i?{THEN}{ELSE}`). This latter is not yet used for
UDFs because all parameters are mandatory. However, it will be very useful for
any custom CAST's datatype.

gmantele committed:
[ADQL] Improve CAST so that it is possible to have a custom target type
+ Improve DefaultUDF and FunctionDef to use as well a FunctionTranslator.

gmantele committed:
[ADQL] Merge DefaultUDF into UserDefinedFunction and Deprecate DefaultUDF.

gmantele committed:
[ADQL] Fix Comparison creation ; it was impossible to compare a POINT to a
column.

gmantele committed:
[ADQL] Fix text position of CENTROID(...)

gmantele committed:
[ADQL] Add `UNION`, `INTERSECT` and `EXCEPT`
WARNING: since this commit, the ADQL-Lib API is different:
         the parser returns an `ADQLSet` instead of an `ADQLQuery`

gmantele committed:
[ADQL] Allow `CAST(NULL AS ...)`

gmantele committed:
[ADQL] Add `COALESCE(...)`

gmantele committed:
[ADQL] Fix recursively wrapped constraints and expressions

gmantele committed:
[ADQL] Support CAST to POINT, CIRCLE and POLYGON.
A custom translation is however needed. This
will be implemented through the TAP configuration
file.

gmantele committed:
[ADQL] Add CIRCLE and POLYGON to the list of supported target data types in CAST

gmantele committed:
[ADQL] Fix COALESCE: only one argument is legal.

gmantele committed:
[ADQL] Fix flag to support extended REGION parameter

gmantele committed:
[ADQL] Allow NULL as valid value expression.

gmantele committed:
[ADQL] Fix geometries syntax with optional coord sys

Supporting NULL in place of a coord sys string
introduced a grammatical ambiguity. This commit
fixes this.

gmantele committed:
[ADQL] Fix translation of CAST with geometries for MySQL

gmantele committed:
[ADQL] Adapt tests to the new geometrical types

gmantele committed:
[ADQL] Add NullValue operand
Fixes #139

gmantele committed:
[ADQL] Accept UNION and INTERSECT in STC-s/REGION

However, no translation and no ADQL object representation
exists for both of them.

mbtaylor committed:
[ADQL] replace #features-adql-geo with #features-adqlgeo (#140)

 [ADQL] replace `#features-adql-geo` with `#features-adqlgeo`

Given the change in the ADQL 2.1 draft at
https://github.com/ivoa-std/ADQL/commit/18a9c19a77c,
LanguageFeature.TYPE_ADQL_GEO should read `#features-adqlgeo` instead
of `#features-adql-geo`.

gmantele committed:
[ADQL] Accept `_` in param name of a UDF declaration

Fixes #141

gmantele committed:
[ADQL] Remove `!` from temporarily accepted col name characters

Fixes #142

olebole commented:
@gmantele Is there any chance to get this merged and released any time soon? I would need it in the Debian package, because new latest Topcat/Stilts releases depend on this, and I would prefer to have a released version of ADQL in Debian instead of an intermediate state of a random branch.

gmantele commented:
> @gmantele Is there any chance to get this merged and released any time soon? I would need it in the Debian package, because new latest Topcat/Stilts releases depend on this, and I would prefer to have a released version of ADQL in Debian instead of an intermediate state of a random branch.

Hi @olebole. Currently, this branch only covers the ADQL library. Some developments are still required to make the TAP library able to work with this new version of the ADQL library. That's why it is not yet merged with `master`. I am currently working on it. I hope it could be ready by the end of the year, but regarding all the work I have also in parallel I cannot give any guarantee. So, I fear it does not match your "any time soon"...sorry for that.

I could however release a beta version of just the ADQL library, although it might change few parts of it until the alpha version. Would it be a better solution for you ou would you prefer to wait for the stable version?

olebole commented:
Thank you for the explanation! A beta version of just ADQL would already be fine for me, since it is much easier to refer the Debian package to it than to a PR on a forked repository. In Debian, we only have the ADQL package (yet), because this is required by @mbtaylor's TOPCAT/STILTS application, and Debian policy is to build everything from source.

tsdower commented:
A beta version of just ADQL would also be useful for the MAST TAP service! We are currently running only the ADQL library in a small microservice wrapper, which validates and translates ADQL to the dialects and spatial query functionality for our individual catalog databases.

gmantele committed:
Initial split into sub-projects

gmantele committed:
Upgrade astroh2 library

gmantele committed:
Ignore build directories

gmantele committed:
Add Gradle wrappers (no more need to install Gradle to build)

gmantele committed:
Make TestSubQueries independent from TAPLib

gmantele committed:
Fix path for the astroh2 library

gmantele committed:
Make TestConfigurableServiceConnection independent of ADQLLib.

gmantele committed:
Fix JavaCC grammar classes generation

gmantele committed:
Simplify TAPLib and UWSLib build configuration

gmantele committed:
Set the common build configuration

gmantele committed:
Set version number in build config for all libraries

gmantele committed:
Use Gradle plugin java-library-distribution
+ Include ADQL examples in classpath (but not JAR)
+ Use STIL from Maven in ADQLLib

gmantele committed:
Write a README for all projects
and update the main README

gmantele committed:
Set up CI to build and test each project

gmantele committed:
Add CI badges in README files

gmantele committed:
Add a license badge in README files

gmantele committed:
Fix a test in ADQLLib and an import in UWSLib

gmantele committed:
Remove former `examples/` directory

gmantele commented:
> Thank you for the explanation! A beta version of just ADQL would already be fine for me, since it is much easier to refer the Debian package to it than to a PR on a forked repository. In Debian, we only have the ADQL package (yet), because this is required by @mbtaylor's TOPCAT/STILTS application, and Debian policy is to build everything from source.

Sorry for answering so late. The beta version of ADQLLib is now available: https://github.com/gmantele/vollt/releases/tag/adqlLib-2.0-beta

This version points toward the tag [adqlLib-2.0-beta](https://github.com/gmantele/vollt/tree/adqlLib-2.0-beta) on the branch [adql2.1](https://github.com/gmantele/vollt/tree/adql2.1). When ready for an alpha release, this branch will be merged with [master](https://github.com/gmantele/vollt/tree/master).

If you need to build the JAR of this library from sources, you can use Gradle provided in the [adql2.1](https://github.com/gmantele/vollt/tree/adql2.1) branch: `./gradlew :ADQLLib:jar`

mbtaylor commented:
Thanks Gregory.

I have updated the version that will be bundled with topcat releases to this adqlLib-2.0-beta version.  I'm not expecting any noticeable change in behaviour since the one I was using previously (8b88a44779) but it's tidier to have a named release.

olebole commented:
Thank you very much! The versio  adql-2.0-beta works fine for me, is not uploaded to Debian, and resolves the issue with missing symbols on starjavas "vo" package.
One minor issue for the future, however:Please do *not* call the next version 2.0-alpha. Usually, alpha is an early pre-release, and beta a later one. And the Debian versioning depends of the correct order and would not allow to upload the -alpha after the -beta.

gmantele committed:
[TAP] Add a draft Docker configuration for TAPLib

gmantele committed:
[ALL] Update GitIgnore rules

gmantele committed:
[ADQL] Fix DB catalog name normalization

Fix #150

gmantele committed:
[ADQL] Fix Javadoc, remove useless code and lightly optimize few elements

gmantele committed:
[ADQL] Fix incorrect ambiguous column name in subqueries

Fixes #158

tsdower commented:
Superceded by https://github.com/spacetelescope/vollt/pull/8
